groups/reconcile.go

```diff
@@ -181,69 +184,176 @@ func main() {
+	log.Println(" ======================= Updates =======================")
+	err = r.reconcileGroups(groupsConfig.Groups)
+	if err != nil {
+		log.Fatal(err)
+	}
+}
+
+// reconciler syncs the actual state of the world with the configuration.
+// It does so by making use of AdminService and GroupService which are mockable
+// interfaces.
+type reconciler struct {
+	adminService AdminService
+	groupService GroupService
+}
+
+func newReconciler(ctx context.Context, clientOption option.ClientOption) (*reconciler, error) {
+	as, err := NewAdminService(ctx, clientOption)
+	if err != nil {
+		return nil, err
+	}
+
+	gs, err := NewGroupService(ctx, clientOption)
+	if err != nil {
+		return nil, err
+	}
+
+	return &reconciler{adminService: as, groupService: gs}, nil
+}
+
+func (r *reconciler) reconcileGroups(groups []GoogleGroup) error {
+	for _, g := range groups {
 		if g.EmailId == "" {
-			log.Fatal(fmt.Sprintf("Group has no email-id: %#v", g))
+			return fmt.Errorf("group has no email-id: %#v", g)
 		}
 
-		err = createOrUpdateGroupIfNecessary(adminService, g.EmailId, g.Name, g.Description)
+		// update the group that is currently being considered.
+		r.adminService.SetGroup(g)
+		r.groupService.SetGroup(g)
+
+		err := r.adminService.CreateOrUpdateGroupIfNescessary()
 		if err != nil {
-			log.Fatal(err)
+			return err
 		}
-		err = updateGroupSettings(groupsSettingsService, g.EmailId, g.Settings)
+		err = r.groupService.UpdateGroupSettings()
 		if err != nil {
-			log.Fatal(err)
+			return err
 		}
-		err = addOrUpdateMembersToGroup(adminService, g.EmailId, g.Owners, "OWNER")
+		err = r.adminService.AddOrUpdateGroupMembers(ownerRole, g.Owners)
 		if err != nil {
-			log.Fatal(err)
+			return err
 		}
-		err = addOrUpdateMembersToGroup(adminService, g.EmailId, g.Managers, "MANAGER")
+		err = r.adminService.AddOrUpdateGroupMembers(managerRole, g.Managers)
 		if err != nil {
-			log.Fatal(err)
+			return err
 		}
-		err = addOrUpdateMembersToGroup(adminService, g.EmailId, g.Members, "MEMBER")
+		err = r.adminService.AddOrUpdateGroupMembers(memberRole, g.Members)
 		if err != nil {
 			log.Println(err)
 		}
+
 		if g.Settings["ReconcileMembers"] == "true" {
 			members := append(g.Owners, g.Managers...)
 			members = append(members, g.Members...)
-			err = removeMembersFromGroup(adminService, g.EmailId, members)
+			err = r.adminService.RemoveMembersFromGroup(members)
 			if err != nil {
-				log.Fatal(err)
+				return err
 			}
 		} else {
 			members := append(g.Owners, g.Managers...)
-			err = removeOwnerOrManagersGroup(adminService, g.EmailId, members)
+			err = r.adminService.RemoveOwnerOrManagersFromGroup(members)
 			if err != nil {
-				log.Fatal(err)
+				return err
 			}
 		}
 	}
-	err = deleteGroupsIfNecessary(adminService)
+	err := r.adminService.DeleteGroupsIfNecessary()
 	if err != nil {
-		log.Fatal(err)
+		return err
+	}
+
+	return nil
+}
+
+func (r *reconciler) printGroupMembersAndSettings() error {
```

Comment:
What would make this easier to review is to leave this function where it was instead of reordering

Make whatever changes you're going to make to the function

Then reorder in a new commit